src/v/archival/archival_policy.cc

```diff
@@ -319,140 +396,18 @@ static ss::future<> get_file_range(
-        auto ix_end = segment->index().find_nearest(end_inclusive.value());
-
-        // NOTE: Index lookup might return an offset which isn't committed yet.
-        // Subsequent call to segment_reader::data_stream will fail in this
-        // case. In order to avoid this we need to make another index lookup
-        // to find a lower offset which is committed.
-        while (ix_end && ix_end->filepos >= fsize) {
-            vlog(archival_log.debug, "The position is not flushed {}", *ix_end);
-            auto lookup_offset = ix_end->offset - model::offset(1);
-            ix_end = segment->index().find_nearest(lookup_offset);
-            vlog(archival_log.debug, "Re-adjusted position {}", *ix_end);
-        }
-
-        size_t scan_from = ix_end ? ix_end->filepos : 0;
-        model::offset fo = ix_end ? ix_end->offset
-                                  : segment->offsets().base_offset;
-        vlog(
-          archival_log.debug,
-          "Segment index lookup returned: {}, scanning from pos {} - offset {}",
-          ix_end,
-          scan_from,
-          fo);
-
-        auto reader_handle = co_await segment->reader().data_stream(
-          scan_from, io_priority);
-        auto ostr = make_null_output_stream();
-        model::timestamp ts = upl.max_timestamp;
-        size_t stop_at = 0;
-        auto res = co_await storage::transform_stream(
-          reader_handle.take_stream(),
-          std::move(ostr),
-          [off_end = end_inclusive.value(), &fo, &ts](
-            model::record_batch_header& hdr) {
-              if (hdr.last_offset() <= off_end) {
-                  // If last offset of the record batch is within the range
-                  // we need to add it to the output stream (to calculate the
-                  // total size).
-                  fo = hdr.last_offset();
-                  // TODO: update max_timestamp
-                  return storage::batch_consumer::consume_result::accept_batch;
-              }
-              ts = hdr.max_timestamp;
-              return storage::batch_consumer::consume_result::stop_parser;
-          });
-        co_await reader_handle.close();
-
-        if (res.has_error()) {
-            vlog(
-              archival_log.error,
-              "Can't read segment file, error: {}",
-              res.error());
-            throw std::system_error(res.error());
-        } else {
-            stop_at = scan_from + res.value();
-            vlog(
-              archival_log.debug,
-              "Scanned {} bytes starting from {}, total {}. Adjusted final "
-              "offset: {}",
-              res.value(),
-              scan_from,
-              stop_at,
-              fo);
-        }
-        upl.final_offset = fo;
-        upl.final_file_offset = stop_at;
-        upl.max_timestamp = ts;
+        auto seek = co_await convert_end_offset_to_file_pos(
+          end_inclusive.value(), segment, upl.max_timestamp, io_priority);
```

Comment:
here? it's hard to tell. better to take segment by value into the coroutine and pass by value here.